guidance/_grammar.py

```diff
@@ -243,6 +243,28 @@ def match_byte(self, byte):
     def max_tokens(self):
         return 1000000000000
 
+class Box(Terminal):
+    """Container to hold a value that is resolved at a later time. This is useful for recursive definitions."""
+    __slots__ = "_value"
+
+    def __init__(self):
+        super().__init__(temperature=-1, capture_name=None)
+        self._resolved = False
+        self._value = None
+
+    @property
+    def value(self):
+        if self._resolved:
+            return self._value
+        else:
+            raise ValueError("Box is not yet resolved")
```

Comment:
TODO: better error messages here?

Reply:
Ultimately, but regular guidance users shouldn't be going here